src/plugins_types.c

```diff
@@ -914,97 +932,113 @@ ly_type_store_binary(const struct ly_ctx *ctx, const struct lysc_type *type, con
+            (('a' <= value[base64_end]) && (value[base64_end] <= 'z')) ||
+            (('0' <= value[base64_end]) && (value[base64_end] <= '9')) ||
+            ('+' == value[base64_end]) ||
+            ('/' == value[base64_end]) ||
+            ('\n' == value[base64_end]))) {
+
+        if ('\n' != value[base64_end]) {
+            base64_count++;
         }
+        base64_end++;
+    }
 
-        for (count = 0, u = start; u <= stop; u++) {
-            if (value[u] == '\n') {
-                /* newline formatting */
-                continue;
-            }
-            count++;
-
-            if (((value[u] < '/') && (value[u] != '+')) ||
-                    ((value[u] > '9') && (value[u] < 'A')) ||
-                    ((value[u] > 'Z') && (value[u] < 'a')) || (value[u] > 'z')) {
-                /* non-encoding characters */
-                if (value[u] == '=') {
-                    /* padding */
-                    if ((u == stop - 1) && (value[stop] == '=')) {
-                        termination = 2;
-                        count++;
-                        u++;
-                    } else if (u == stop) {
-                        termination = 1;
-                    }
-                }
-                if (!termination) {
-                    /* error */
-                    rc = asprintf(&errmsg, "Invalid Base64 character (%c).", value[u]);
-                    goto error;
-                }
-            }
+    /* find end of padding */
+    base64_terminated = 0;
+    while (((base64_end < value_len) && (base64_terminated < 2)) &&
+            /* check padding on end of string */
+            (('=' == value[base64_end]) ||
+            ('\n' == value[base64_end]))) {
+
+        if ('\n' != value[base64_end]) {
+            base64_terminated++;
+        }
+        base64_end++;
+    }
+
+    /* check if value is valid base64 value */
+    if (value_end != base64_end) {
+        ret = LY_EVALID;
+        if (asprintf(&errmsg, "Invalid Base64 character (%c).", value[base64_end]) < 0) {
+            errmsg = NULL;
         }
+        goto error;
     }
 
-finish:
-    if (count & 3) {
+    if ((base64_count + base64_terminated) & 3) {
         /* base64 length must be multiple of 4 chars */
+        ret = LY_EVALID;
         errmsg = strdup("Base64 encoded value length must be divisible by 4.");
         goto error;
     }
 
-    /* length of the encoded string */
+    /* check if value meets the type requirments */
     if (type_bin->length) {
-        char buf[LY_NUMBER_MAXLEN];
-        /* get correct length based on base64 encoding rules */
-        uint64_t len = ((count / 4) * 3) - termination;
-        snprintf(buf, LY_NUMBER_MAXLEN, "%" PRIu64, len);
-        LY_CHECK_RET(ly_type_validate_range(LY_TYPE_BINARY, type_bin->length, len, buf, err));
+        const unsigned value_length = ((base64_count + base64_terminated) / 4) * 3 - base64_terminated;
```

Comment:
Please do not use `unsigned` or `int unsigned` or even `unsigned int`, they have variable length which can cause problems. Just use `uint32_t` instead.